edb/server/protocol/pg_ext.pyx

```diff
@@ -1329,6 +1333,33 @@ cdef class PgConnection(frontend.FrontendConnection):
             self.debug_print("extended_query", actions)
         return actions, None
 
+    def check_capabilities(
+        self,
+        query_unit,
+    ):
+        if query_unit.required_permissions:
+            is_superuser, permissions = self.get_permissions()
+            if not is_superuser:
+                for perm in query_unit.required_permissions:
+                    if perm not in permissions:
+                        missing = sorted(
+                            set(query_unit.required_permissions)
+                            - set(permissions)
+                        )
+                        plural = 's' if len(missing) > 1 else ''
+                        raise errors.DisabledCapabilityError(
+                            f'role {self.username} does not have required '
+                            f'permission{plural}: {", ".join(missing)}'
+                        )
```

Comment:
Another option would be to add a SQL_SESSION_CONFIG to the `Capability` enum.
I don't know if I have a particularly strong opinion

Reply:
Oh, functions are something we need to check too, if we support calling edgeql functions from SQL, which I can't remember if we do